### PR TITLE
Prevent redundant transitions

### PR DESCRIPTION
**Pull request by anticorrelator**
Here we add a simple orchestration rule preventing redundant transitions. Among other things, this will prevent multiple Agents from orchestrating an active run in competing and inconsistent ways.

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a `fix`, `feature`, `enhancement`, `docs`, `collections`, or `maintenance` label categorizing the change.

### CODE REVIEW

File: src/prefect/orion/orchestration/core_policy.py [hunk 1]
```diff
@@ -407,3 +410,40 @@ async def before_transition(
...
+        context: OrchestrationContext,
+    ) -> None:
+        initial_state_type = initial_state.type if initial_state else None
+        proposed_state_type = proposed_state.type if proposed_state else None
+        if (
+            self.STATE_PROGRESS[proposed_state_type]
+            <= self.STATE_PROGRESS[initial_state_type]
+        ):
+            await self.abort_transition(
+                reason=f"This run cannot to the {proposed_state.type} state."
```
Comment (jlowin) [inline src/prefect/orion/orchestration/core_policy.py:448]: I think a word is missing here - "transition"?

File: src/prefect/orion/orchestration/core_policy.py [hunk 2]
```diff
@@ -407,3 +410,40 @@ async def before_transition(
...
+    """
+    Prevents redundant transitions.
+
+    Under normal operation, this rule prevents the "backwards" progress of a run. This
+    rule will also help prevent multiple agents from attempting to orchestrate a run by
+    preventing transitions into the same state type. If any of these disallowed
+    transitions are attempted, this rule will abort the transition.
+    """
+
+    STATE_PROGRESS = {
```
Comment (jlowin) [inline src/prefect/orion/orchestration/core_policy.py:425]: This is really neat

File: src/prefect/orion/orchestration/core_policy.py [hunk 3]
```diff
@@ -34,6 +35,7 @@ class CoreFlowPolicy(BaseOrchestrationPolicy):
     def priority():
         return [
             PreventTransitionsFromTerminalStates,
+            PreventBackwardsTransitions,
```
Comment (jlowin) [inline src/prefect/orion/orchestration/core_policy.py:38]: Where is this class? Maybe a rename?

**Review by jlowin — CHANGES_REQUESTED** [review]
Great stuff @anticorrelator and a win for the policy design to see how compact it is!

I don't think it's critical but I'd love to see an API-level unit test that proves this from a user's point of view (rather than the engine's point of view) by calling `set_flow_run_state(RUNNING)` twice in a row.

File: src/prefect/orion/orchestration/core_policy.py (hunk 3 above)
Comment (anticorrelator) [reply src/prefect/orion/orchestration/core_policy.py:38]: oops! yes, this was a rename

**Comment by madkinsz** [comment]
Just wondering, should we use REJECT and provide the current state allowing the client to decide if it should retry or give up? Or is this always going to be a case where the run is already being handled elsewhere and consequently we should force an ABORT?

**Comment by anticorrelator** [comment]
I believe that if these transitions should never be allowed, the only reason why a client might attempt them is if we're in a bad state; I prefer aborting here instead of trying to solve for the cases where we might be able to recover

File: src/prefect/engine.py [hunk 4]
```diff
@@ -1169,11 +1169,22 @@ async def orchestrate_task_run(
...
-            task_run_id=task_run.id,
-        )
+        if task_run.state.is_pending():
+            # if orchestrating a run already in a pending state, force orchestration to
+            # update the state name
+            return await propose_state(
+                client,
+                Pending(name="NotReady", message=str(upstream_exc)),
+                task_run_id=task_run.id,
+                force=True,
```
Comment (madkinsz) [inline src/prefect/engine.py:1179]: Can we just do `force=task_run.state.is_pending()` with that comment for clarity?

**Comment by jlowin** [comment]
I agree, REJECT means "you asked to transition to X but you transitioned to Y instead" which isn't the semantic here

**Review by jlowin — APPROVED** [review]

**Pull request merged**